dynix.3.2.0/src/doc/man1/vwidth.1
.\" $Copyright: $
.\" Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991
.\" Sequent Computer Systems, Inc.   All rights reserved.
.\"  
.\" This software is furnished under a license and may be used
.\" only in accordance with the terms of that license and with the
.\" inclusion of the above copyright notice.   This software may not
.\" be provided or otherwise made available to, or used by, any
.\" other person.  No title to or ownership of the software is
.\" hereby transferred.
.\"
.\" This software is furnished under a license and may be used
.\" only in accordance with the terms of that license and with the
.\" inclusion of the above copyright notice.  This software may not
.\" be provided or otherwise made available to, or used by, any
.\" other person.  No title to or ownership of the software is
.\" hereby transferred.
...
.V= $Header: vwidth.1 1.9 1991/04/15 19:41:20 $
.TH VWIDTH 1 "\*(V)" "3BSD/DYNIX"
.SH NAME
/usr/lib/vwidth \- make troff width table for a font
.SH SYNOPSIS
.B vwidth
fontfile pointsize > ft\f2xx\fP.c
.br
.B cc -c ft\f2xx\fP.c
.br
mv ft\f2xx\fP.o /usr/lib/font/ft\f2xx\fP

.B vwidth \-ascii
fontfile pointsize > /usr/lib/font/ft\f2xx\fP
.SH DESCRIPTION
.I Vwidth
translates from the width information stored in the vfont style
format to the format expected by
.I troff .
Non-DYNIX versions of
.I troff
want an object file in
.IR a.out (5)
format.
(This fact does not seem to be documented anywhere.)
.I Troff
should look directly in the font file but it doesn't.
.PP
.I Vwidth
should be used after editing a font with
.IR fed (1).
It is not necessary to use
.I vwidth
unless you have made a change
that would affect the width tables.
Such changes include numerically editing the width field,
adding a new character,
and moving or copying a character to a new position.
It is
.I not
always necessary to use
.I vwidth
if the physical width of the glyph
(e.g. the number of columns in the bit matrix) has changed, but
if it has changed much the logical width should probably be changed
and
.I vwidth
run.
.PP
If the
.B \-ascii
option is not selected,
.I vwidth
produces a C program on its standard output.
This program should be run through the C compiler and
the object (that is, the .o file) saved.
The resulting file should be placed in
.I /usr/lib/font
in the file
.RI ft xx
where
.II xx
is a one or two letter code that is the logical (internal to troff)
font name.
This name can be found by looking in the file
.RI /usr/lib/fontinfo/ fname *
where
.I fname
is the external name of the font.
.PP
If the
.B \-ascii
option is selected,
.I vwidth
produces an ASCII-readable version of the width file,
for use by the DYNIX
.IR troff .
The resulting file should be placed directly in
.I /usr/lib/font
without further processing.
.SH "FILES"
/usr/lib/vwidth
.SH "SEE ALSO"
fed(1),
vfont(5),
troff(1),
vtroff(1)
.SH BUGS
Produces the C file using obsolete syntax
that the portable C compiler complains about.


dynix.3.2.0/src/ucb/lpr/filters/vwidth.c
/* $Copyright:	$
 * Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990 
 * Sequent Computer Systems, Inc.   All rights reserved.
 *  
 * This software is furnished under a license and may be used
 * only in accordance with the terms of that license and with the
 * inclusion of the above copyright notice.   This software may not
 * be provided or otherwise made available to, or used by, any
 * other person.  No title to or ownership of the software is
 * hereby transferred.
 */

#ifndef lint
static char rcsid[] = "$Header: vwidth.c 2.0 86/01/28 $";
#endif

/*
 *	Creates a width table for troff from a versatec font for a
 *		normal font.
 *	Usage: width font [ point_size ]
 *		where font is the file name of the versatec font and
 *		point_size is it point size.
 *	If the point size is omitted it is taken from the suffix of
 *	the font name, as bocklin.14 => 14 point.
 *	It is better to use as large a point size font as possible 
 *	to avoid round off.
 */

#include <stdio.h>

struct wtable {
	char	charloc;
	char	*name;
} wtable[] = {
	'\214',	"space",
	'!',	"!",
	'"',	"\"",
	'#',	"#",
	'$',	"$",
	'%',	"%",
	'&',	"&",
	'\'',	"'",
	'(',	"(",
	')',	")",
	'*',	"*",
	'+',	"+",
	',',	",",
	'-',	"- hyphen",
	'.',	".",
	'/',	"/",
	'0',	"0",
	'1',	"1",
	'2',	"2",
	'3',	"3",
	'4',	"4",
	'5',	"5",
	'6',	"6",
	'7',	"7",
	'8',	"8",
	'9',	"9",
	':',	":",
	';',	";",
	'<',	"<",
	'=',	"=",
	'>',	">",
	'?',	"?",
	'@',	"@",
	'A',	"A",
	'B',	"B",
	'C',	"C",
	'D',	"D",
	'E',	"E",
	'F',	"F",
	'G',	"G",
	'H',	"H",
	'I',	"I",
	'J',	"J",
	'K',	"K",
	'L',	"L",
	'M',	"M",
	'N',	"N",
	'O',	"O",
	'P',	"P",
	'Q',	"Q",
	'R',	"R",
	'S',	"S",
	'T',	"T",
	'U',	"U",
	'V',	"V",
	'W',	"W",
	'X',	"X",
	'Y',	"Y",
	'Z',	"Z",
	'[',	"[",
	'\\',	"\\",
	']',	"]",
	'^',	"^",
	'_',	"_",
	'\`',	"\`",
	'a',	"a",
	'b',	"b",
	'c',	"c",
	'd',	"d",
	'e',	"e",
	'f',	"f",
	'g',	"g",
	'h',	"h",
	'i',	"i",
	'j',	"j",
	'k',	"k",
	'l',	"l",
	'm',	"m",
	'n',	"n",
	'o',	"o",
	'p',	"p",
	'q',	"q",
	'r',	"r",
	's',	"s",
	't',	"t",
	'u',	"u",
	'v',	"v",
	'w',	"w",
	'x',	"x",
	'y',	"y",
	'z',	"z",
	'{',	"{",
	'|',	"|",
	'}',	"}",
	'~',	"~",
	'\206',	"narrow space",
	'-',	"hyphen",
	'\07',	"bullet",
	'\010',	"square",
	'\06',	"3/4 em dash",
	'\05',	"rule",
	'\021',	"1/4",
	'\022',	"1/2",
	'\023',	"3/4",
	'\04',	"minus",
	'\01',	"fi",
	'\02',	"fl",
	'\03',	"ff",
	'\011',	"ffi",
	'\012',	"ffl",
	'\013',	"degree",
	'\014',	"dagger",
	'\200',	"section (unimplem)",
	'\015',	"foot mark",
	'\200',	"acute acc (unimplem)",
	'\200',	"grave acc (unimplem)",
	'\200',	"underrule (unimplem)",
	'\200',	"slash (unimplem)",
	'\203',	"half narrow space",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\200',	"null",
	'\017',	"registered",
	'\016',	"copyright",
	'\200',	"null",
	'\020',	"cent",
	0,	0
};

struct desc {
	short	addr;
	short	nbytes;
	char	up;
	char	down;
	char	left;
	char	right;
	short	width;
} desc[256];

int	aflag;			/* portable ascii format output */

main(argc, argv)
	int argc;
	char *argv[];
{
	register int cl;
	register esc;
	register w;
	int i, psize;
	int fd, high;

	if (argc != 4 && argc != 3 && argc != 2) {
		printf("Usage: vwidth [-ascii] font [pointsize] > font.c\n");
		exit(1);
	}
	if (!strcmp(argv[1], "-ascii"))
		++argv, --argc, ++aflag;
	fd = open(argv[1], 0);
	if (argc == 3)
		psize = atoi(argv[2]);
	else {
		char *p;
		for (p = argv[1]; *p && *p != '.'; p++)
			;
		if (*p == 0) {
			psize = 10;
			fprintf(stderr, "Assuming %d point\n", psize);
		} else
			psize = atoi(p+1);
	}
	lseek(fd, 10L, 0);
	read(fd, desc, sizeof desc);
	high = desc['a'].up+1;
	if (!aflag)
		printf("char XXw[256-32] = {\n");
	for (i = 0; wtable[i].charloc != 0; i++) {
		cl = wtable[i].charloc & 0377;
		if (cl & 0200)
			w = cl & 0177;
		else
			w = desc[cl].width*(54./25.)*(6./psize)+.5;
		esc = 0;
		if ((cl >= '0' && cl <= '9') || (cl >= 'A' && cl <= 'Z') ||
		    (cl >= 'a' && cl <= 'z')) {
			if (desc[cl].up > high)
				esc |= 0200;
			if (desc[cl].down > 0)
				esc |= 0100;
		}
		if (!aflag) {
			if (esc)
			  printf("%d+0%o,\t/* %s */\n", w, esc, wtable[i].name);
			else
			  printf("%d,\t\t/* %s */\n", w, wtable[i].name);
		} else
			printf("%3d /*%d*/\n", (w+esc) & 0xff, i);
	}
	if (!aflag)
		printf("};\n");
	else
		while (i < (256-32))
			printf("%3d /*%d*/\n", 0, i++);
}
